README.rst
=========================
 Google ngram downloader
=========================

.. image:: https://travis-ci.org/dimazest/google-ngram-downloader.png?branch=master
    :target: https://travis-ci.org/dimazest/google-ngram-downloader

.. image:: https://coveralls.io/repos/dimazest/google-ngram-downloader/badge.png?branch=master
    :target: https://coveralls.io/r/dimazest/google-ngram-downloader?branch=master

.. image:: https://zenodo.org/badge/4321/dimazest/google-ngram-downloader.png
    :target: http://dx.doi.org/10.5281/zenodo.11884
    :alt: Zenodo doi.

`The Google Books Ngram Viewer dataset`__ is a freely available resource under
a `Creative Commons Attribution 3.0 Unported License`__ which provides ngram
counts over books scanned by Google.

__ http://storage.googleapis.com/books/ngrams/books/datasetsv2.html
__ http://creativecommons.org/licenses/by/3.0/

The data is so big, that storing it is almost impossible. However, sometimes
you need an aggregate data over the dataset. For example to build a
co-occurrence matrix.

This package provides an iterator over the dataset stored at Google. It
decompresses the data on the fly and provides you the access to the underlying
data.

Features
========

* Download ngrams of various length and languages.
* Access to part of ngrams, e.g. ones that start with an 'a'.

Installation
============

::

    pip install google-ngram-downloader


The command line tool
=====================

It also provides a simple command line tool to download the ngrams called
`google-ngram-downloader`. Refer to the help to see available actions::

    google-ngram-downloader help
    usage: google-ngram-downloader <command> [options]

    commands:

     cooccurrence  Write the cooccurrence frequencies of a word and its contexts.
     download      Download The Google Books Ngram Viewer dataset version 20120701.
     help          Show help for a given help topic or a help overview.
     readline      Print the raw content.


Example use of the API
======================

>>> from google_ngram_downloader import readline_google_store
>>>
>>> fname, url, records = next(readline_google_store(ngram_len=5))
>>> fname
'googlebooks-eng-all-5gram-20120701-0.gz'
>>> url
'http://storage.googleapis.com/books/ngrams/books/googlebooks-eng-all-5gram-20120701-0.gz'
>>> next(records)
Record(ngram=u'0 " A most useful', year=1860, match_count=1, volume_count=1)


CHANGES.rst
Changes
=======

Version 4.0.1
-------------

* Citation information.
* Tests are correctly packaged for a release.

Version 4.0.0
-------------

* Added 'indices' keyword. Thanks to neocortex.
* Added 'language' flat. Thanks to Ray Powell (rpowellgit).

Version 3.1.1
-------------

* Non-unique contexts are taken into account inside of an ngram.

Version 3.1
-----------

* The ``cooccurrence`` command does not perform any ngram modification.

Version 3.0
-----------

* `download`, `readile` and `cooccurrence` subcommands.
* `readline_google_store` transforms lines to `Record` in several processes.


test/conftest.py
import pytest

from google_ngram_downloader.util import Record


@pytest.fixture
def unigram_indices():
    return set(
        '0 1 2 3 4 5 6 7 8 9 a b c d e f g h i j k l m n o other p pos '
        'punctuation q r s t u v w x y z'.split(' ')
    )


@pytest.fixture
def bigrams_indices():
    return set(
        '0 1 2 3 4 5 6 7 8 9 _ADJ_ _ADP_ _ADV_ _CONJ_ _DET_ _NOUN_ _NUM_ _PRON_ '
        '_PRT_ _VERB_ a_ aa ab ac ad ae af ag ah ai aj ak al am an ao ap aq ar '
        'as at au av aw ax ay az b_ ba bb bc bd be bf bg bh bi bj bk bl bm bn bo '
        'bp bq br bs bt bu bv bw bx by bz c_ ca cb cc cd ce cf cg ch ci cj ck cl '
        'cm cn co cp cq cr cs ct cu cv cw cx cy cz d_ da db dc dd de df dg dh di '
        'dj dk dl dm dn do dp dq dr ds dt du dv dw dx dy dz e_ ea eb ec ed ee ef '
        'eg eh ei ej ek el em en eo ep eq er es et eu ev ew ex ey ez f_ fa fb fc '
        'fd fe ff fg fh fi fj fk fl fm fn fo fp fq fr fs ft fu fv fw fx fy fz g_ '
        'ga gb gc gd ge gf gg gh gi gj gk gl gm gn go gp gq gr gs gt gu gv gw gx '
        'gy gz h_ ha hb hc hd he hf hg hh hi hj hk hl hm hn ho hp hq hr hs ht hu '
        'hv hw hx hy hz i_ ia ib ic id ie if ig ih ii ij ik il im in io ip iq ir '
        'is it iu iv iw ix iy iz j_ ja jb jc jd je jf jg jh ji jj jk jl jm jn jo '
        'jp jq jr js jt ju jv jw jx jy jz k_ ka kb kc kd ke kf kg kh ki kj kk kl '
        'km kn ko kp kq kr ks kt ku kv kw kx ky kz l_ la lb lc ld le lf lg lh li '
        'lj lk ll lm ln lo lp lq lr ls lt lu lv lw lx ly lz m_ ma mb mc md me mf '
        'mg mh mi mj mk ml mm mn mo mp mq mr ms mt mu mv mw mx my mz n_ na nb nc '
        'nd ne nf ng nh ni nj nk nl nm nn no np nq nr ns nt nu nv nw nx ny nz o_ '
        'oa ob oc od oe of og oh oi oj ok ol om on oo op oq or os ot other ou ov '
        'ow ox oy oz p_ pa pb pc pd pe pf pg ph pi pj pk pl pm pn po pp pq pr ps '
        'pt pu punctuation pv pw px py pz q_ qa qb qc qd qe qf qg qh qi qj qk ql '
        'qm qn qo qp qq qr qs qt qu qv qw qx qy qz r_ ra rb rc rd re rf rg rh ri '
        'rj rk rl rm rn ro rp rq rr rs rt ru rv rw rx ry rz s_ sa sb sc sd se sf '
        'sg sh si sj sk sl sm sn so sp sq sr ss st su sv sw sx sy sz t_ ta tb tc '
        'td te tf tg th ti tj tk tl tm tn to tp tq tr ts tt tu tv tw tx ty tz u_ '
        'ua ub uc ud ue uf ug uh ui uj uk ul um un uo up uq ur us ut uu uv uw ux '
        'uy uz v_ va vb vc vd ve vf vg vh vi vj vk vl vm vn vo vp vq vr vs vt vu '
        'vv vw vx vy vz w_ wa wb wc wd we wf wg wh wi wj wk wl wm wn wo wp wq wr '
        'ws wt wu wv ww wx wy wz x_ xa xb xc xd xe xf xg xh xi xj xk xl xm xn xo '
        'xp xq xr xs xt xu xv xw xx xy xz y_ ya yb yc yd ye yf yg yh yi yj yk yl '
        'ym yn yo yp yq yr ys yt yu yv yw yx yy yz z_ za zb zc zd ze zf zg zh zi '
        'zj zk zl zm zn zo zp zq zr zs zt zu zv zw zx zy zz'.split(' ')
    )


@pytest.fixture
def records():
    return (
        Record('a BB z', 1987, 10, 1),
        Record('a BB z', 1988, 100, 1),
        Record('a BB z', 1989, 1000, 1),
        Record('a ABCDEFG z', 1989, 222, 1),
    )


@pytest.fixture
def ngram():
    return 'aa ab BB zz zz'


test/test_main.py
# -*- coding: utf8 -*-
import gzip
import zlib
from contextlib import contextmanager

from requests import Session

from google_ngram_downloader.__main__ import download, cooccurrence, readline
from google_ngram_downloader import util

import pytest


@pytest.fixture
def compressobj():
    return zlib.compressobj()


@pytest.fixture
def urls():
    return []


@pytest.fixture
def data():
    return (
        b'analysis is often described as\t1991\t1\t1\n',
        b'analysis', b' is often described as\t1992\t2\t1\n',
        b'analysis', b' is often', b' described as\t1993\t3\t1\n',
        b'c1 c2 WORD c3 c4\t1987\t100\t2\n',
        b'aa aa REPETITION aa aa\t1999\t10\t1\n',
        b'\xd1\x8e \xd1\x83 UNICODE \xd1\x83 \xd1\x8e\t2002\t23\t3\n',
    )


@pytest.fixture
def compressed_data(compressobj, data):
    return tuple(map(compressobj.compress, data)) + (compressobj.flush(), )


@pytest.fixture(autouse=True)
def fake_request(urls, compressed_data, monkeypatch):
    def mocked_get(self, url, **kwargs):
        urls.append(url)

        class FakeRequest:
            def iter_content(self, chunk_size):
                return iter(compressed_data)

            status_code = 200

        return FakeRequest()

    monkeypatch.setattr(Session, 'get', mocked_get)


@pytest.mark.parametrize(
    ('verbose', 'err_len'),
    (
        (False, 1),
        (True, 725),
    ),
)
def test_download(capsys, tmpdir, verbose, err_len, urls):
    download.command(
        '-o {tmpdir} -n 2 {verbose}'
        ''.format(
            tmpdir=tmpdir,
            verbose='-v' if verbose else '',
        ).split()
    )

    assert len(urls) == len(tmpdir.listdir()) == 724

    out, err = capsys.readouterr()
    assert not out
    assert len(err.split('\n')) == err_len


def test_readline(capsys, tmpdir):
    readline.command([])

    out, err = capsys.readouterr()

    assert len(out) == 149868
    assert not err


@pytest.mark.parametrize(
    'verbose',
    (
        (False, ),
        (True, ),
    ),
)
def test_cooccurrence(tmpdir, monkeypatch, verbose):
    objects = []

    def modked_open(obj, *args, **kwargs):
        @contextmanager
        def _open(*args, **kwargs):
            class MockedGzipFile(object):
                def writelines(self, lines):
                    objects.extend(list(lines))
            yield MockedGzipFile()

        return _open()

    monkeypatch.setattr(util, 'get_indices', lambda ngram_len: ['a'])

    cooccurrence.command(
        '-o {tmpdir} -n 5 --records-in-file 3 {verbose}'
        ''.format(
            tmpdir=tmpdir,
            verbose='-v' if verbose else '',
        ).split()
    )

    def read(f_name):
        with gzip.open(str(f_name), mode='rb') as f:
            return sorted(f.read().decode('utf-8').split(u'\n'))

    result_one, result_two = map(read, tmpdir.listdir(sort=True))

    assert result_one == [
        u'',
        u'often\tanalysis\t6',
        u'often\tas\t6',
        u'often\tdescribed\t6',
        u'often\tis\t6',
    ]

    assert result_two == [
        u'',
        u'REPETITION\taa\t40',
        u'UNICODE\tу\t46',
        u'UNICODE\tю\t46',
        u'WORD\tc1\t100',
        u'WORD\tc2\t100',
        u'WORD\tc3\t100',
        u'WORD\tc4\t100',
    ]


test/test_util.py
from google_ngram_downloader.util import (
    get_indices,
    ngram_to_cooc,
    count_coccurrence,
)

import pytest


def test_get_indices_1grams(unigram_indices):
    indices = list(get_indices(1))
    assert len(set(indices)) == len(indices)

    assert set(indices) == unigram_indices


def test_get_indices_manygrams(bigrams_indices):
    indices = list(get_indices(2))
    assert len(set(indices)) == len(indices)

    assert set(indices) == bigrams_indices


def test_get_indices_5grams(bigrams_indices):
    """Check that there is no "qk" index for 5grams."""
    indices = list(get_indices(5))
    assert len(indices) == len(set(bigrams_indices)) - 1

    assert set(indices) == (bigrams_indices - set(['qk']))


@pytest.mark.parametrize(
    ('ngram', 'expected_result', 'index'),
    (
        (
            'aa_DET ab BB zz yz',
            (
                ((0, 5), 3),
                ((0, 2), 3),
                ((0, 3), 3),
                ((0, 4), 3),
            ),
            {'BB': 0, 'aa': 1, 'ab': 2, 'zz': 3, 'yz': 4, 'aa_DET': 5},
        ),
        (
            'aa yz BB yz yz',
            (
                ((0, 1), 3),
                ((0, 2), 3),
                ((0, 2), 3),
                ((0, 2), 3),
            ),
            {'BB': 0, 'aa': 1},
        ),
    ),
)
def test_ngrams_to_cooc(ngram, expected_result, index):
    result = ngram_to_cooc(ngram, 3, index)
    assert result == expected_result


def test_count_coccurrence(records):
    index = {}
    assert count_coccurrence(records, index) == {
        (index['BB'], index['a']): 1110,
        (index['BB'], index['z']): 1110,
        (index['ABCDEFG'], index['a']): 222,
        (index['ABCDEFG'], index['z']): 222,
    }


google_ngram_downloader/__main__.py
import gzip
import sys
from collections import OrderedDict
from itertools import islice

from opster import Dispatcher
from py.path import local

from .util import iter_google_store, readline_google_store, count_coccurrence


dispatcher = Dispatcher()
command = dispatcher.command


@command()
def download(
    ngram_len=('n', 1, 'The length of ngrams to be downloaded.'),
    output=('o', 'downloads/google_ngrams/{ngram_len}', 'The destination folder for downloaded files.'),
    verbose=('v', False, 'Be verbose.'),
    lang=(
        'l',
        'eng',
        'Language. [eng|eng-us|eng-gb|eng-fiction|chi-sim|fre|ger|heb|ita|rus|spa]',
    ),
):
    """Download The Google Books Ngram Viewer dataset version 20120701."""
    output = local(output.format(ngram_len=ngram_len))
    output.ensure_dir()

    for fname, url, request in iter_google_store(ngram_len, verbose=verbose, lang=lang):
        with output.join(fname).open('wb') as f:
            for num, chunk in enumerate(request.iter_content(1024)):
                if verbose and not divmod(num, 1024)[1]:
                    sys.stderr.write('.')
                    sys.stderr.flush()
                f.write(chunk)


@command()
def cooccurrence(
    ngram_len=('n', 2, 'The length of ngrams to be downloaded.'),
    output=('o', 'downloads/google_ngrams/{ngram_len}_cooccurrence', 'The destination folder for downloaded files.'),
    verbose=('v', False, 'Be verbose.'),
    rewrite=('r', False, 'Always rewrite existing files.'),
    records_in_file=(
        '',
        50000000,
        'The number of records to be read from the Google store to store in a .json.gz file.',
    ),
    lang=(
        'l',
        'eng',
        'Language. [eng|eng-us|eng-gb|eng-fiction|chi-sim|fre|ger|heb|ita|rus|spa]',
    ),
):
    """Write the cooccurrence frequencies of a word and its contexts."""
    assert ngram_len > 1
    output_dir = local(output.format(ngram_len=ngram_len))
    output_dir.ensure_dir()

    for fname, _, all_records in readline_google_store(ngram_len, lang=lang,  verbose=verbose):
        postfix = 0
        while (True):
            records = islice(all_records, records_in_file)
            output_file = output_dir.join(
                '{fname}_{postfix}.gz'.format(
                    fname=fname,
                    postfix=postfix,
                )
            )

            if not rewrite and output_file.check():
                if verbose:
                    print('Skipping {} and the rest...'.format(output_file))
                break

            index = OrderedDict()
            cooccurrence = count_coccurrence(records, index)

            if not cooccurrence:
                break

            id2word = list(index)
            items = (u'{}\t{}\t{}\n'.format(id2word[i], id2word[c], str(v)) for (i, c), v in cooccurrence.items())

            with gzip.open(str(output_file), 'wb') as f:
                if verbose:
                    print('Writing {}'.format(output_file))
                for item in items:
                    f.write(item.encode('utf8'))

            postfix += 1


@command()
def readline(
    ngram_len=('n', 2, 'The length of ngrams to be downloaded.'),
    lang=(
        'l',
        'eng',
        'Language. [eng|eng-us|eng-gb|eng-fiction|chi-sim|fre|ger|heb|ita|rus|spa]',
    ),
):
    """Print the raw content."""

    for _, _, records in readline_google_store(ngram_len, lang=lang):
        for record in records:
            print(u'{ngram}\t{year}\t{match_count}\t{volume_count}'.format(**record._asdict()))


google_ngram_downloader/util.py
import collections
import sys
import zlib
from itertools import product, chain, groupby
from string import ascii_lowercase, digits

import requests


URL_TEMPLATE = 'http://storage.googleapis.com/books/ngrams/books/{}'
# URL_TEMPLATE = 'http://localhost:8001/{}'

FILE_TEMPLATE = 'googlebooks-{lang}-all-{ngram_len}gram-{version}-{index}.gz'


Record = collections.namedtuple('Record', 'ngram year match_count volume_count')


def readline_google_store(ngram_len, lang='eng', indices=None, chunk_size=1024 ** 2, verbose=False):
    """Iterate over the data in the Google ngram collectioin.

        :param int ngram_len: the length of ngrams to be streamed.
        :param str lang: the langueage of the ngrams.
        :param iter indices: the file indices to be downloaded.
        :param int chunk_size: the size the chunks of raw compressed data.
        :param bool verbose: if `True`, then the debug information is shown to `sys.stderr`.

        :returns: a iterator over triples `(fname, url, records)`

    """

    for fname, url, request in iter_google_store(ngram_len, verbose=verbose, lang=lang, indices=indices):
        dec = zlib.decompressobj(32 + zlib.MAX_WBITS)

        def lines():
            last = b''
            compressed_chunks = request.iter_content(chunk_size=chunk_size)

            for i, compressed_chunk in enumerate(compressed_chunks):
                chunk = dec.decompress(compressed_chunk)

                lines = (last + chunk).split(b'\n')
                lines, last = lines[:-1], lines[-1]

                for line in lines:
                    line = line.decode('utf-8')
                    data = line.split('\t')
                    assert len(data) == 4
                    ngram = data[0]
                    other = map(int, data[1:])
                    yield Record(ngram, *other)

            assert not last

        yield fname, url, lines()


def ngram_to_cooc(ngram, count, index):
    ngram = ngram.split()

    middle_index = len(ngram) // 2
    item = ngram[middle_index]
    context = ngram[:middle_index] + ngram[middle_index + 1:]

    item_id = word_to_id(item, index)
    context_ids = (word_to_id(c, index) for c in context)

    return tuple((p, count) for p in product([item_id], context_ids))


def word_to_id(word, index):
    try:
        return index[word]
    except KeyError:
        id_ = len(index)
        index[word] = id_
        return id_


def count_coccurrence(records, index):
    grouped_records = groupby(records, key=lambda r: r.ngram)
    ngram_counts = ((ngram, sum(r.match_count for r in records)) for ngram, records in grouped_records)
    cooc = (ngram_to_cooc(ngram, count, index) for ngram, count in ngram_counts)

    counter = collections.Counter()
    for item, count in chain.from_iterable(cooc):
        counter[item] += count

    return counter


def iter_google_store(ngram_len, lang="eng", indices=None, verbose=False):
    """Iterate over the collection files stored at Google.

    :param int ngram_len: the length of ngrams to be streamed.
    :param str lang: the langueage of the ngrams.
    :param iter indices: the file indices to be downloaded.
    :param bool verbose: if `True`, then the debug information is shown to `sys.stderr`.

    """
    version = '20120701'
    session = requests.Session()

    indices = get_indices(ngram_len) if indices is None else indices

    for index in indices:
        fname = FILE_TEMPLATE.format(
            lang=lang,
            ngram_len=ngram_len,
            version=version,
            index=index,
        )

        url = URL_TEMPLATE.format(fname)

        if verbose:
            sys.stderr.write(
                'Downloading {url} '
                ''.format(
                    url=url,
                ),
            )
            sys.stderr.flush()

        request = session.get(url, stream=True)
        assert request.status_code == 200

        yield fname, url, request

        if verbose:
            sys.stderr.write('\n')


def get_indices(ngram_len):
    """Generate the file indeces depening on the ngram length, based on version 20120701.

    For 1grams it is::

        0 1 2 3 4 5 6 7 8 9 a b c d e f g h i j k l m n o other p pos
        punctuation q r s t u v w x y z

    For others::

        0 1 2 3 4 5 6 7 8 9 _ADJ_ _ADP_ _ADV_ _CONJ_ _DET_ _NOUN_ _NUM_ _PRON_
        _PRT_ _VERB_ a_ aa ab ac ad ae af ag ah ai aj ak al am an ao ap aq ar
        as at au av aw ax ay az b_ ba bb bc bd be bf bg bh bi bj bk bl bm bn bo
        bp bq br bs bt bu bv bw bx by bz c_ ca cb cc cd ce cf cg ch ci cj ck cl
        cm cn co cp cq cr cs ct cu cv cw cx cy cz d_ da db dc dd de df dg dh di
        dj dk dl dm dn do dp dq dr ds dt du dv dw dx dy dz e_ ea eb ec ed ee ef
        eg eh ei ej ek el em en eo ep eq er es et eu ev ew ex ey ez f_ fa fb fc
        fd fe ff fg fh fi fj fk fl fm fn fo fp fq fr fs ft fu fv fw fx fy fz g_
        ga gb gc gd ge gf gg gh gi gj gk gl gm gn go gp gq gr gs gt gu gv gw gx
        gy gz h_ ha hb hc hd he hf hg hh hi hj hk hl hm hn ho hp hq hr hs ht hu
        hv hw hx hy hz i_ ia ib ic id ie if ig ih ii ij ik il im in io ip iq ir
        is it iu iv iw ix iy iz j_ ja jb jc jd je jf jg jh ji jj jk jl jm jn jo
        jp jq jr js jt ju jv jw jx jy jz k_ ka kb kc kd ke kf kg kh ki kj kk kl
        km kn ko kp kq kr ks kt ku kv kw kx ky kz l_ la lb lc ld le lf lg lh li
        lj lk ll lm ln lo lp lq lr ls lt lu lv lw lx ly lz m_ ma mb mc md me mf
        mg mh mi mj mk ml mm mn mo mp mq mr ms mt mu mv mw mx my mz n_ na nb nc
        nd ne nf ng nh ni nj nk nl nm nn no np nq nr ns nt nu nv nw nx ny nz o_
        oa ob oc od oe of og oh oi oj ok ol om on oo op oq or os ot other ou ov
        ow ox oy oz p_ pa pb pc pd pe pf pg ph pi pj pk pl pm pn po pp pq pr ps
        pt pu punctuation pv pw px py pz q_ qa qb qc qd qe qf qg qh qi qj qk ql
        qm qn qo qp qq qr qs qt qu qv qw qx qy qz r_ ra rb rc rd re rf rg rh ri
        rj rk rl rm rn ro rp rq rr rs rt ru rv rw rx ry rz s_ sa sb sc sd se sf
        sg sh si sj sk sl sm sn so sp sq sr ss st su sv sw sx sy sz t_ ta tb tc
        td te tf tg th ti tj tk tl tm tn to tp tq tr ts tt tu tv tw tx ty tz u_
        ua ub uc ud ue uf ug uh ui uj uk ul um un uo up uq ur us ut uu uv uw ux
        uy uz v_ va vb vc vd ve vf vg vh vi vj vk vl vm vn vo vp vq vr vs vt vu
        vv vw vx vy vz w_ wa wb wc wd we wf wg wh wi wj wk wl wm wn wo wp wq wr
        ws wt wu wv ww wx wy wz x_ xa xb xc xd xe xf xg xh xi xj xk xl xm xn xo
        xp xq xr xs xt xu xv xw xx xy xz y_ ya yb yc yd ye yf yg yh yi yj yk yl
        ym yn yo yp yq yr ys yt yu yv yw yx yy yz z_ za zb zc zd ze zf zg zh zi
        zj zk zl zm zn zo zp zq zr zs zt zu zv zw zx zy zz

    Nothe, there is not index "qk" for 5grams.

    See http://storage.googleapis.com/books/ngrams/books/datasetsv2.html for
    more details.

    """
    other_indices = ('other', 'punctuation')

    if ngram_len == 1:
        letter_indices = ascii_lowercase
        other_indices += 'pos',

    else:
        letter_indices = ((''.join(i) for i in product(ascii_lowercase, ascii_lowercase + '_')))
        if ngram_len == 5:
            letter_indices = (l for l in letter_indices if l != 'qk')

        other_indices += (
            '_ADJ_',
            '_ADP_',
            '_ADV_',
            '_CONJ_',
            '_DET_',
            '_NOUN_',
            '_NUM_',
            '_PRON_',
            '_PRT_',
            '_VERB_',
        )

    return chain(digits, letter_indices, other_indices)
